### PR TITLE
Implement support for APA102 led strips

### PR DESCRIPTION
**Pull request by heyjuvi**
I made an implementation for the APA102 led chips. I took the NeoPixel implementation as a basis, but I'm not sure if it is correct to implement it only for the esp8266 and not more generally.
If i did anything wrong or can improve something, please let me know!

### CODE REVIEW

**Comment by pfalcon** [comment]
Thanks for the patch. It may take some time before we'll be able to look into this. In the meantime, there was another similar patch posted: #1950, it would be nice if you guys cooperated to produce one well tested module.

**Comment by robertfoss** [comment]
@misterdanb, Apparently there are two APA102 PRs floating about.
Personally I don't have a preference about which one is merged, but if you'd like, feel free to include any snippets of #1950 into this PR.

**Comment by robertfoss** [comment]
@misterdanb I was hoping to get a hold of you the see if we could combine our efforts.

**Comment by misterdanb** [comment]
@robertfoss Oh, sorry, didn't look into this pull request for quite a while (was a little bit busy with other stuff). I will have a look at this in the next week, i hope. Actually some parts of my code are not that nice, because I put everything in one function (as GPIO writes did not work within helper functions and I had no idea why, very weird behaviour or I do understand somthing very wrong), so I'm very fine with taking some of your code! I will test it on my devices as soon as possible :)…

**Comment by misterdanb** [comment]
@robertfoss I finally found some time to restructure code, and took some of yours for that. I hope things got somehow cleaner.

**Comment by robertfoss** [comment]
@misterdanb looks good to me!

**Comment by robertfoss** [comment]
@pfalcon This PR is ready to be merged.

**Comment by pfalcon** [comment]
@robertfoss: Can you please rebase onto the latest master?

File: docs/esp8266/quickref.rst [hunk 1]
```diff
@@ -274,6 +274,26 @@ For low-level driving of a NeoPixel::
...
+    data = Pin(13, Pin.OUT)      # set GPIO13 to output to drive the data
+    apa = APA102(clock, data, 8) # create APA102 driver on the clock and the data pin for 8 pixels
+    apa[0] = (255, 255, 255, 31) # set the first pixel to white with a maximum brightness of 31
+    apa.write()                  # write data to all pixels
+    r, g, b, brightness = apa[0] # get first pixel colour
+
+For low-level driving of an APA102::
+
+    import esp
+    esp.neopixel_write(clock_pin, data_pin, rgbi_buf)
```
Comment (mcauser) [inline docs/esp8266/quickref.rst]: Should this be `esp.apa102_write()`?

File: esp8266/scripts/apa102.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+from esp import apa102_write
+
+class APA102:
+    def __init__(self, clock_pin, data_pin, n):
+        self.clock_pin = clock_pin
+        self.data_pin = data_pin
+        self.n = n
+        self.buf = bytearray(n * 4)
+
```
Comment (mcauser) [inline esp8266/scripts/apa102.py]: Do you need to set them as output pins here?
`self.clock_pin.init(pin.OUT)`
`self.data_pin.init(pin.OUT)`

File: docs/esp8266/quickref.rst (hunk 1 above)
Comment (misterdanb) [inline docs/esp8266/quickref.rst]: Yes you're right. Thanks!

File: esp8266/scripts/apa102.py (hunk 2 above)
Comment (misterdanb) [inline esp8266/scripts/apa102.py]: And you're also right with this one, i think. :)

**Comment by misterdanb** [comment]
Okay, so after fixing the mentioned issues, reviewing the whole pull request again and testing the code again with a 60 leds strip, it should be fine now. But please have a look at it again. too, of course!

**Comment by pfalcon** [comment]
@misterdanb : MicroPython uses rebase-based workflow, so all changes will need to be rebased onto the current master, and all merge commits removed. https://github.com/micropython/micropython/wiki/DevelWorkflow

**Comment by misterdanb** [comment]
Rebased everything, sorry for missing that workflow.

**Comment by pfalcon** [comment]
@misterdanb : Thanks, and getting there. In MicroPython, all (or almost all files) should have a copyright/licensing header, see e.g. 
https://github.com/micropython/micropython/blob/master/esp8266/esponewire.c . Would you add it to espapa102.c/espapa102.h (with your and/or @robertfoss' copyrights)? The reason not to do that would be if you copied that code from somewhere, in which case it should have comments about its source and its licensing/copyright (and licensing should be compatible with MicroPython's of course). Thanks.

**Comment by misterdanb** [comment]
Added licensing information as seen in similar files. @robertfoss Are you okay with the MIT license?

**Comment by misterdanb** [comment]
Okay, since no one answers, I think MIT license is fine. ^^

**Comment by pfalcon** [comment]
@misterdanb: I assume @robertfoss is aware that MicroPython is MIT-licensed and was OK with that if he submitted his pull request. At least that's how we handle (as most of other community-backed projects here on github): we don't ask for contributor agreement or explicit sign-off, and assume that act of submitting pull request constitutes agreement with the license of the project.

And sorry for this still being backlogged, I'm actually looking into merging this today.

**Comment by pfalcon** [comment]
Ok, I tried to see how to squash this down to 2-3 commits, but ended up just squashing altogether. Thanks for contribution @misterdanb , @robertfoss and following thru the review process!

**Comment by robertfoss** [comment]
I 100% ok with MIT licensing my contributions. Thanks for verifying it
though.
On May 19, 2016 15:24, "Paul Sokolovsky" <notifications@github.com> wrote:

@misterdanb <https://github.com/misterdanb>: I assume @robertfoss
<https://github.com/robertfoss> is aware that MicroPython is MIT-licensed
and was OK with that if he submitted his pull request. At least that's how
we handle (as most of other community-backed projects here on github): we
don't ask for contributor agreement or explicit sign-off, and assume that
act of submitting pull request constitutes agreement with the license of
the project.

And sorry for this still being backlogged, I'm actually looking into
merging this today.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/micropython/micropython/pull/1941#issuecomment-220425080>

**Comment by mattytrentini** [comment]
I'm looking into helping create an [APA102 driver for the ESP32 port](https://github.com/micropython/micropython-esp32/issues/121). I can see that the ESP8266 implementation is specific to that chipset but I guess I'm trying to understand why a bit-banged approach was chosen over using SPI?

I'm probably oversimplifying but it seems like a pure python APA102 driver that _used_ `machine.SPI` would work across all ports and be optimised for ports that had (hardware-enabled) SPI...

**Comment by dpgeorge** [comment]
> I'm trying to understand why a bit-banged approach was chosen over using SPI?

I'm guessing that at the time it wasn't clear to us that SPI could be used... but @misterdanb would be the best person to answer that.

> I'm probably oversimplifying but it seems like a pure python APA102 driver that used machine.SPI would work across all ports

Yes that sounds like it would work, and it would use less code because of the re-use of the SPI class.  Probably the main drawback is that SPI currently requires SCK/MOSI/MISO so that would tie up an extra pin that's not needed (MISO).  So one way thing to do would be extend the SPI class to allow it to work without MISO or MOSI (and possibly even SCK?).  I would say that most MCUs could support this in their hardware SPI block, just by not configuring the relevant GPIO to be controlled by the SPI.

**Comment by misterdanb** [comment]
That is totally an option! I used SPI to drive the APA102 in another (Arduino based) project myself. The reason, why i chose to use the bit-bangend approach is that you are not limited to use the SPi, if you need it for something else (or even want to mux those pins otherwise).
I am currently not so much aware of the ESP32's hardware, so I cannot speak for this one.

But even though it would be nice to have a more efficient way of driving the APAs. That this then comes for free for other MCUs wasn't on my mind at that time.

**Comment by mattytrentini** [comment]
Got it, thanks @dpgeorge and @misterdanb!

I'll try to spend some time putting together a pure Python APA102 driver based on SPI and see what it looks like. 

I understand the compromise about the extra pin currently needed for SPI but agree with @dpgeorge's suggestion that the SPI class should be more flexible in allowing pins to not be allocated (even though some hardware may not allow that). However, my opinion is that even if the SPI class _isn't_ extended in this way having an efficient, portable implementation seems worth the potential loss of a pin. 

I'll also take a closer look at existing solutions and see if there's value in porting something. 

The thought did cross my mind that we could wrap [FastLED](http://fastled.io/) but I'm not sure that'll be worth the effort...

**Pull request closed without merging**